### ISSUE TITLE
Make rancher.internal search path order configurable

### ISSUE THREAD

**Issue by DanielDent** [rancher/rancher#3572]
Currently, if a rancher container is started on a host with a search path of "example.com", containers will be launched with a search path of "example.com rancher.internal".

If foo.example.com exists, bar service calling foo will connect to foo.example.com rather than foo.rancher.internal. 

I'm not 100% sure on implementation details, but I'm also concerned that this may lead to unnecessary latency & DNS queries leaking.

I think having rancher.internal as the first choice in the DNS search path is a more sensible default.

**Comment by vincent99** [comment]
The search paths on hosts are generally useless on a public cloud but often exist for a reason in an enterprise environment.. putting Rancher last is a lot safer than allowing any user to create a service that accidentally breaks an unqualified name other users/code expect to work based on one of the existing search paths.

Really if this is something you're concerned about you should just always be using the fully-qualified `foo.rancher.internal` to get Rancher records.  Then it's fully under your control and unambiguous.

**Comment by DanielDent** [comment]
That's an interesting perspective. We actually do set custom search paths on our hosts as part of our configuration management.

I opened this issue this up having experienced the opposite of what you describe: a name was created in a domain in the search path which ended up breaking a rancher-managed service.

**Comment by vincent99** [comment]
Right, so either way one camp is unhappy and both should just use FQDNs :smile:

This isn't the first time someone has brought it up though, so probably a setting is in order...

**Comment by DanielDent** [comment]
FQDNs are great except that using a search path is somtimes a nice way to discover a particular instance of a service, i.e. 'in the same datacenter', or 'belongs to the same customer', and various other approaches.

A setting to make it configurable seems like a great approach to me.

**Comment by vincent99** [comment]
But the Rancher ones are always just += `.rancher.internal`, and we conveniently add ourselves last so it doesn't conflict with entries of those approaches if you always make the Rancher ones fully-qualified :wink:

**Comment by m4ce** [comment]
I actually opened the same request here https://github.com/rancher/rancher/issues/3460 a few days back.

**Comment by scanon** [comment]
So has this been taken off the table?  This can lead to a many standard catalog recipes not working and it can be very hard to diagnose.  It seems there should be an option to specify whether or not to include the search from the host system or whether to prepend or append it.

**Comment by stefanlasiewski** [comment]
Due to this issue, the "CREATE A MULTI-CONTAINER APPLICATION" section in the [Quick Start Guide](http://docs.rancher.com/rancher/v1.2/en/quick-start-guide/) doesn't work for me.

The guide says:

> In the **Service Links**, add the database service and provide the name _mysql_.

However, in my case, the Wordpress containers cannot find the mysql container:

```
rancher1 # docker logs --tail 10 stack1_mywordpress_1
Warning: mysqli::mysqli(): (HY000/2002): php_network_getaddresses: getaddrinfo failed: Name or service not known in - on line 19
MySQL Connection Error: (2002) php_network_getaddresses: getaddrinfo failed: Name or service not known
```

As a workaround, if I manually log into the container and hardcode the IP address for the mysql container, the Wordpress container can reach the mysql container:

```
rancher1 # docker exec -it stack1_mywordpress_1 sh
# echo "10.42.11.115 mysql mysql.stack1.rancher.internal" >> /etc/hosts && cat /etc/hosts
...
172.17.0.3      76fc1781bda4
10.42.11.115 mysql mysql.stack1.rancher.internal
# exit
```

```
rancher1 # docker logs --tail 10 stack1_mywordpress_1
[Tue Oct 25 21:51:00.014896 2016] [mpm_prefork:notice] [pid 1] AH00163: Apache/2.4.10 (Debian) PHP/5.6.27 configured -- resuming normal operations
```

Another workaround is to hardcode the Wordpress container with an environment variable like 'WORDPRESS_DB_HOST=mysql.stack1:3306'; or perhaps modifying the "Networking: Search Domains" parameter.

However, neither of these workarounds are sustainable for very long.

**Comment by fabiorauber** [comment]
I agree with @scanon, this breaks standard docker functionality. A stack of services that runs fine in my desktop with docker-compose (with the same search path) does not run in my rancher environment, because one of the containers has the same name of a service within my enterprise network. In my opinion, rancher.internal should come first.

**Comment by bseng** [comment]
I want to ask is this implemented in v1.4.1? 

I've another issue while now the rancher.internal comes before searches on the host in resolv.conf, the search string is longer than 256 characters and the domain append at the end cannot be search. 

I need to manually change the resolv.conf to have 2 search lines to fix the issue.

i.e.

> search instance.rancher.interal rancher.internal
> search example.org
> nameserver ...



ref http://superuser.com/questions/570082/in-etc-resolv-conf-what-exactly-does-the-search-configuration-option-do

**Comment by superseb** [comment]
Currently, we place it first and can be configured using the label `io.rancher.container.dns.priority: service_last` (http://rancher.com/docs/rancher/latest/en/cattle/labels/). Please file a new issue if you are running into anything.
